store/src/db/migrations.rs

```diff
@@ -1,71 +1,90 @@
-        ) STRICT, WITHOUT ROWID;
-
-        CREATE TABLE
-            accounts
-        (
-            account_id INTEGER NOT NULL,
-            account_hash BLOB NOT NULL,
-            block_num INTEGER NOT NULL,
-            details BLOB,
-
-            PRIMARY KEY (account_id),
-            CONSTRAINT fk_block_num FOREIGN KEY (block_num) REFERENCES block_headers (block_num),
-            CONSTRAINT accounts_block_num_is_u32 CHECK (block_num BETWEEN 0 AND 0xFFFFFFFF)
-        ) STRICT, WITHOUT ROWID;
-
-        CREATE TABLE
-            nullifiers
-        (
-            nullifier BLOB NOT NULL,
-            nullifier_prefix INTEGER NOT NULL,
-            block_num INTEGER NOT NULL,
-
-            PRIMARY KEY (nullifier),
-            CONSTRAINT fk_block_num FOREIGN KEY (block_num) REFERENCES block_headers (block_num),
-            CONSTRAINT nullifiers_nullifier_is_digest CHECK (length(nullifier) = 32),
-            CONSTRAINT nullifiers_nullifier_prefix_is_u16 CHECK (nullifier_prefix BETWEEN 0 AND 0xFFFF),
-            CONSTRAINT nullifiers_block_num_is_u32 CHECK (block_num BETWEEN 0 AND 0xFFFFFFFF)
-        ) STRICT, WITHOUT ROWID;
-        ",
-    )])
-});
+use rusqlite::Connection;
+use rusqlite_migration::{Migrations, SchemaVersion, M};
+use tracing::{debug, info, instrument};
+
+use crate::{db::settings::Settings, errors::DatabaseError, COMPONENT};
+
+type Hash = Blake3Digest<20>;
+
+const MIGRATION_SCRIPTS: [&str; 1] = [include_str!("migrations/001-init.sql")];
+static MIGRATION_HASHES: Lazy<Vec<Hash>> = Lazy::new(compute_migration_hashes);
+static MIGRATIONS: Lazy<Migrations> = Lazy::new(prepare_migrations);
+
+fn up(s: &'static str) -> M<'static> {
+    M::up(s).foreign_key_check()
+}
+
+const DB_MIGRATION_HASH_FIELD: &str = "db-migration-hash";
+
+#[instrument(target = "miden-store", skip_all, err)]
+pub fn apply_migrations(conn: &mut Connection) -> super::Result<()> {
+    let version_before = MIGRATIONS.current_version(conn)?;
+
+    info!(target: COMPONENT, version_before = %version_before, "Running database migrations");
+
+    if let SchemaVersion::Inside(ver) = version_before {
+        if !Settings::exists(conn)? {
+            return Err(DatabaseError::UnsupportedDatabaseVersion);
+        }
+
+        let expected_hash = MIGRATION_HASHES[ver.get() - 1].to_bytes();
+        let actual_hash = Settings::get_value(conn, DB_MIGRATION_HASH_FIELD)?;
+
+        debug!(
+            target: COMPONENT,
+            expected_hash = %hex::encode(&expected_hash),
+            actual_hash = ?actual_hash.as_ref().map(hex::encode),
+            "Comparing migration hashes",
+        );
+
+        if actual_hash != Some(expected_hash) {
+            return Err(DatabaseError::UnsupportedDatabaseVersion);
+        }
+    }
+
+    MIGRATIONS.to_latest(conn).map_err(DatabaseError::MigrationError)?;
+
+    if version_before != MIGRATIONS.current_version(conn)? {
```

Comment:
This doesn't seem to be catching errors when the number of migrations is the same. I did the following:

- run the node to create the DB
- opened the DB via the CLI, and ran `DROP TABLE accounts; CREATE TABLE accounts (a int);` to simulate conflicting migrations
- ran the node, and it didn't detect the issue, so the code ran, and I got an error inside the code `SQLite error: no such column: account_id in SELECT account_id, account_hash FROM accounts ORDER BY block_num ASC; at offset 7`

I think we should always check the hash.

Reply:
This solution is not intended to check changing of the database structure by the users outside of our migration mechanism. If you think, this worth it, I can redesign my solution, but it would require to apply all migrations to in-memory database and compare hashes with hash calculated for current database structure. For hash calculation we would need to use pragmas: `table_list`, `table_info` (or even `table_xinfo`), `index_list`, `index_xinfo`, `foreign_key_list`, and calculate hashes of the returned results. And I don't know, if it's possible to extract constraint checks. I think, this would be too complex for the problem we try to solve.

A much simpler solution might be just storing schema version after migration in `settings` table and comparing it on the next start. If database schema was changed outside of our migrations, we can raise an error and fail to start.

Reply:
> This solution is not intended to check changing of the database structure by the users outside of our migration mechanism

Agreed, and that is not what I was trying to propose. I did the manually edit as a test, to show that changing the migration files is not caught by this change, unless I'm missing something.

> A much simpler solution might be just storing schema version after migration in settings table and comparing it on the next start. If database structure was changed outside of our migrations, the schema version will be changed, so we can raise an error and fail to start.

Ye, that is what I was trying to get at, this `if` statement is preventing the comparison against the settings table.

Reply:
> Yes, that is what I was trying to get at, this `if` statement is preventing the comparison against the settings table.

I probably misunderstand you, but this `if` is after migration is happened to the latest supported version and all checks were successful.

Reply:
> I probably misunderstand you, but if statement you hilighted is positioned after migration is happened to the latest supported version and all checks succeeded.

Ah, my bad. This is indeed the wrong `if`. I haven't looked deeper why the hash check didn't catch the issue.